src/main/java/school/hei/haapi/repository/dao/CourseDao.java

```diff
@@ -0,0 +1,146 @@
+package school.hei.haapi.repository.dao;
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.persistence.EntityManager;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Join;
+import javax.persistence.criteria.JoinType;
+import javax.persistence.criteria.Order;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import lombok.AllArgsConstructor;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.query.QueryUtils;
+import org.springframework.stereotype.Repository;
+import school.hei.haapi.endpoint.rest.model.CourseStatus;
+import school.hei.haapi.model.Course;
+import school.hei.haapi.model.StudentCourse;
+import school.hei.haapi.model.User;
+
+import static school.hei.haapi.endpoint.rest.model.CourseStatus.LINKED;
+import static school.hei.haapi.endpoint.rest.model.CourseStatus.UNLINKED;
+
+@Repository
+@AllArgsConstructor
+public class CourseDao {
+  private final EntityManager entityManager;
+
+  public List<Course> findByCriteria(String code, String name, Integer credits,
+                                     String teacherFirstName, String teacherLastName,
+                                     String creditsOrder, String codeOrder,
+                                     Pageable pageable) {
+    CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<Course> query = builder.createQuery(Course.class);
+    Root<Course> root = query.from(Course.class);
+    Join<Course, User> teacher = root.join("mainTeacher");
+
+    List<Predicate> predicates = new ArrayList<>();
+
+    predicates.add(
+        builder.or(
+            builder.like(builder.lower(root.get("code")), "%" + code + "%"),
+            builder.like(root.get("code"), "%" + code + "%")
+        )
+    );
+
+    if (name != null) {
```

Comment:
Because here we check if the name attribute is not null, that means the default value must stay null not "" as you did in the CourseController line 31 and under. 

Make all necessary changes in the course Controller then.